File: server/src/commands/permission/cmd_permission.c
/************************************************************************
*            Atrinik, a Multiplayer Online Role Playing Game            *
*                                                                       *
*    Copyright (C) 2009-2011 Alex Tokar and Atrinik Development Team    *
*                                                                       *
* Fork from Daimonin (Massive Multiplayer Online Role Playing Game)     *
* and Crossfire (Multiplayer game for X-windows).                       *
*                                                                       *
* This program is free software; you can redistribute it and/or modify  *
* it under the terms of the GNU General Public License as published by  *
* the Free Software Foundation; either version 2 of the License, or     *
* (at your option) any later version.                                   *
*                                                                       *
* This program is distributed in the hope that it will be useful,       *
* but WITHOUT ANY WARRANTY; without even the implied warranty of        *
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
* GNU General Public License for more details.                          *
*                                                                       *
* You should have received a copy of the GNU General Public License     *
* along with this program; if not, write to the Free Software           *
* Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.             *
*                                                                       *
* The author can be reached at admin@atrinik.org                        *
************************************************************************/

/**
 * @file
 * Implements the /cmd_permission command.
 *
 * @author Alex Tokar */

#include <global.h>

/** @copydoc command_func */
void command_cmd_permission(object *op, const char *command, char *params)
{
	char word[MAX_BUF];
	size_t pos;
	player *pl;
	int i;

	pos = 0;

	if (!string_get_word(params, &pos, word, sizeof(word)))
	{
		draw_info(COLOR_RED, op, "Usage: /cmd_permission <player> <add|remove|list> [command]");
		return;
	}

	pl = find_player(word);

	if (!pl)
	{
		draw_info(COLOR_WHITE, op, "No such player.");
		return;
	}

	string_get_word(params, &pos, word, sizeof(word));
	params = player_sanitize_input(params + pos);

	if (strcmp(word, "add") == 0)
	{
		if (!params)
		{
			return;
		}

		for (i = 0; i < pl->num_cmd_permissions; i++)
		{
			if (pl->cmd_permissions[i] && !strcmp(pl->cmd_permissions[i], params))
			{
				draw_info_format(COLOR_RED, op, "%s already has permission for /%s.", pl->ob->name, params);
				return;
			}
		}

		pl->num_cmd_permissions++;
		pl->cmd_permissions = realloc(pl->cmd_permissions, sizeof(char *) * pl->num_cmd_permissions);
		pl->cmd_permissions[pl->num_cmd_permissions - 1] = strdup(params);
		draw_info_format(COLOR_GREEN, op, "%s has been granted permission for /%s.", pl->ob->name, params);
	}
	else if (strcmp(word, "remove") == 0)
	{
		if (!params)
		{
			return;
		}

		for (i = 0; i < pl->num_cmd_permissions; i++)
		{
			if (pl->cmd_permissions[i] && !strcmp(pl->cmd_permissions[i], params))
			{
				FREE_AND_NULL_PTR(pl->cmd_permissions[i]);
				draw_info_format(COLOR_GREEN, op, "%s has had permission for /%s command removed.", pl->ob->name, params);
				return;
			}
		}

		draw_info_format(COLOR_RED, op, "%s does not have permission for /%s.", pl->ob->name, params);
	}
	else if (strcmp(word, "list") == 0)
	{
		if (pl->cmd_permissions)
		{
			draw_info_format(COLOR_WHITE, op, "%s has permissions for the following commands:\n", pl->ob->name);

			for (i = 0; i < pl->num_cmd_permissions; i++)
			{
				if (pl->cmd_permissions[i])
				{
					draw_info_format(COLOR_WHITE, op, "/%s", pl->cmd_permissions[i]);
				}
			}
		}
		else
		{
			draw_info_format(COLOR_WHITE, op, "%s has no command permissions.", pl->ob->name);
		}
	}
}

File: maps/python/commands/cmd_permission.py
## @file
## Implements the /cmd_permission command.

import re
from Interface import Interface
from Markup import markup_escape, markup_unescape

def main():
    match = re.match(r"((?:\")(.+)(?:\")|([^ ]+))"
            "( (removeall|(add|remove) ([^ ]+)|add))?", msg)

    if not match:
        activator.Controller().DrawInfo(
                "Usage: /cmd_permission <[]\"]player name[]\"]>",
                color = COLOR_WHITE)
        return

    player = match.group(2) or match.group(3)
    action = match.group(6) or match.group(5)
    cmd_permission = match.group(7)

    pl = FindPlayer(player)

    if not pl:
        activator.Controller().DrawInfo("No such player.", color = COLOR_RED)
        return

    inf = Interface(activator, pl)

    if cmd_permission:
        cmd_permission = markup_unescape(cmd_permission)

    if action == "add":
        if cmd_permission:
            if not cmd_permission in pl.Controller().cmd_permissions:
                pl.Controller().cmd_permissions.append(cmd_permission)
            else:
                inf.add_msg("{pl.name} already has that permission.",
                        color = COLOR_RED, pl = pl)
        else:
            inf.set_text_input(prepend = "/cmd_permission {} add ".format(
                    pl.name))
    elif action == "remove":
        try:
            pl.Controller().cmd_permissions.remove(cmd_permission)
        except ValueError:
            inf.add_msg("{pl.name} does not have that permission.",
                    color = COLOR_RED, pl = pl)
    elif action == "removeall":
        pl.Controller().cmd_permissions.clear()

    inf.add_msg("{pl.name} has the following permissions:\n", pl = pl)

    for tmp in pl.Controller().cmd_permissions:
        if tmp:
            inf.add_msg("\n    {perm} [y=-2][a=:/cmd_permission "
                    "{pl.name} remove {perm}]x[/a]",
                    perm = markup_escape(tmp), newline = False, pl = pl)

    inf.add_link("Add permission",
            dest = "/cmd_permission {} add".format(pl.name))
    inf.add_link("Remove all permissions",
            dest = "/cmd_permission {} removeall".format(pl.name))
    inf.finish()

main()
